_source/_posts/2023-07-27-spring-graphql-react.adoc

```diff
@@ -0,0 +1,1857 @@
+---
+layout: blog_post
+title: "How to build a GraphQL API with Spring for GraphQL"
+author: jimena-garbarino
+by: contractor
+communities: [security,java,javascript]
+description: "A step-by-step guide for building a secured GraphQL API with Spring Boot and Auth0 authentication on React"
+tags: [java, javascript]
+tweets:
+- ""
+- ""
+- ""
+image:
+type: awareness
+---
+:page-liquid:
+:toc: macro
+:experimental:
+
+
+GraphQL is a query language for APIs and a runtime that allows the API consumer to get exactly the required information instead of the server exclusively controlling the contents of the response. While some REST API implementations require loading the references of a resource from multiple URLs, GraphQL APIs can follow references between related objects and return them in a single request. This step-by-step guide demonstrates how to build a GraphQL API with Spring Boot and Spring for GraphQL, for querying a sample dataset of related companies, persons, and properties, seeded to a Neo4j database. It also demonstrates how to build a React client with Next.js and MUI Datagrid to consume the API. Both the client and the server are secured with Auth0 authentication using Okta Spring Boot Starter for the server and Auth0 React SDK for the client.
+
+image::{% asset_path blog/spring-graphql-react/logos.png %}[alt=Spring, GraphQL and React,width=700,align=center]
+
+[NOTE]
+====
+.This example was created with the following tools and services:
+- https://docs.npmjs.com/downloading-and-installing-node-js-and-npm[Node.js v18.16.1]
+- https://docs.npmjs.com/downloading-and-installing-node-js-and-npm[npm 9.5.1]
+- https://jdk.java.net/java-se-ri/17[Java OpenJDK 17]
+- https://docs.docker.com/desktop/[Docker 24.0.2]
+- https://auth0.com/signup[Auth0 account]
+- https://github.com/auth0/auth0-cli#installation[Auth0 CLI 1.0.0]
+- https://httpie.io/[HTTPie 3.2.2]
+- https://nextjs.org/[Next.js 13.4.19]
+====
+
+toc::[]
+
+== Build a GraphQL API with Spring for GraphQL
+
+The resource server is a Spring Boot web application that exposes a GraphQL API with Spring for GraphQL. The API allows querying a Neo4j database containing information about companies and their related owners and properties, using Spring Data Neo4j. The data was obtained from a Neo4j https://neo4j.com/graphgists/35a813ba-ea10-4165-9065-84f8802cbae8/[use case example].
+
+Create the application with Spring Initializr and HTTPie:
+
+[source,shell]
+----
+https start.spring.io/starter.zip \
+  bootVersion==3.1.2 \
+  language==java \
+  packaging==jar \
+  javaVersion==17 \
```

Comment:
These are all defaults. I don't think we need to include them, unless you want to be explicit.

Reply:
I would keep javaVersion explicit, and any other version setting.